README.md
# Flower
[![Build Status](https://travis-ci.org/CodeSteak/Flower.svg?branch=master)](https://travis-ci.org/CodeSteak/Flower)
[![Hex.pm](https://img.shields.io/hexpm/v/flower.svg)](https://hex.pm/packages/flower)


This is an implementation of __Bloom Filters for Elixir__. It uses __NIFs__ written __in Rust__ for better performance, since Bloom Filters rely highly on mutability.

#### What are Bloom Filter?
__TL;DR__: *Huge amount of data __➜__ small Bloom Filter __:__ Was X not in Huge amount of data?*

For more about this topic consider, checking out:
* [Bloom Filters by Jason Davis](https://www.jasondavies.com/bloomfilter/)
* [Bloom Filters, Mining of Massive Datasets, Stanford University on Youtube](https://www.youtube.com/watch?v=qBTdukbzc78)

## Installation

The package can be installed by adding `flower` to your list of dependencies in `mix.exs`:

```elixir
def deps do
  [
    {:flower, "~> 0.1.4"},
  ]
end
```

Also, you need to have __Rust__ installed for development, since this uses NIFs.

Docs can be found at [https://hexdocs.pm/flower](https://hexdocs.pm/flower).

Documentation can be generated with [ExDoc](https://github.com/elixir-lang/ex_doc).

## Usage
There is also a small walk through in the [Example Section](#example)
```elixir
alias Flower.Bloom, as: Bloom
```
##### Create new Bloom Filter
```elixir
expected_number_of_unique_elements = 1000
# With a size atom
iex> filter = Bloom.new(:"8 KB", expected_number_of_unique_elements)
...
# Or by a maximum byte size:
iex> filter = Bloom.new_by_byte_size(8 * 1024, expected_number_of_unique_elements)
...
# Or by a bit address length:
iex> filter = Bloom.new(16, expected_number_of_unique_elements)
...
```
| Bit Address Length|Size Atom| Bit Address Length|Size Atom| Bit Address Length|Size Atom|
|-------:|---------:|-------:|---------:|-------:|---------:|
|        |          | __13__ | 1 KB     | __23__ | 1 MB     |
|        |          | __14__ | 2 KB     | __24__ | 2 MB     |
|        |          | __15__ | 4 KB     | __25__ | 4 MB     |
|  __6__ | 8 Byte   | __16__ | 8 KB     | __26__ | 8 MB     |
|  __7__ | 16 Byte  | __17__ | 16 KB    | __27__ | 16 MB    |
|  __8__ | 32 Byte  | __18__ | 32 KB    | __28__ | 32 MB    |
|  __9__ | 64 Byte  | __19__ | 64 KB    | __29__ | 64 MB    |
| __10__ | 128 Byte | __20__ | 128 KB   | __30__ | 128 MB   |
| __11__ | 256 Byte | __21__ | 256 KB   | __31__ | 256 MB   |
| __12__ | 512 Byte | __22__ | 512 KB   | __32__ | 512 MB   |

##### Insert Elements
```elixir
iex> Bloom.insert(filter, 42)
:ok
iex> Bloom.insert(filter, [1,2,{:atom, <<1,2,3,4>>}])
:ok
iex> Bloom.insert(filter, "Hello!")
:ok
```

##### Check for Elements
```elixir
iex> Bloom.has_not?(filter, "Hello!")
false
iex> Bloom.has?(filter, [1,2,{:atom, <<1,2,3,4>>}])
true
iex> Bloom.has?(filter, :atom)
false
# `has?` is always the opposite of `has_not?`.
iex> Bloom.has?(filter, 42) != Bloom.has_not?(filter, 42)
true
```

|Was actually inserted?|                  has?  |                    has_not?  |
|:--------------------:|:----------------------:|:----------------------------:|
| yes                  | yes                    |                       no     |
| no                   | most of the times: no  |   most of the times: yes     |


##### Write To Disk

```elixir
filename = "prime_filter.bloom"
file = File.stream!(filename, [:delayed_write, :binary], 8096)

(Bloom.stream(filter) |> Stream.into(file) |> Stream.run)
```

##### Read From Disk

```elixir
filename = "prime_filter.bloom"
file = File.stream!(filename, [:read_ahead, :binary], 8096)

{:ok, new_filter} = Bloom.from_stream(file)
```

##### Funky Stuff
```elixir
iex> Bloom.estimate_count(filter)
3
iex> Bloom.false_positive_probability(filter)
3.2348227494719115e-28
# This number is only that small, because we have just 3 elements in 8 KBs.
```
## <a name="example"></a>Example
### Prime Numbers
Checking if a number is not a prime and below `100_000`.
```elixir
# helper module
defmodule Step do
    def throught(from, to, step, func) when from < to do
        func.(from)
        throught(from+step, to, step, func)
    end
    def throught(_,_,_,_), do: :ok
end && :ok

# alias so we need to write less
alias Flower.Bloom, as: Bloom


# Select appropriate size.
# We will put about 100_000 elements in.
# 64 KB = 512 KBit
# 512 KBit / 100_000 ~= 5.25 Bits per element.
# Meeh. Let's see...
non_primes = Bloom.new(:"64 KB", 100_000)
# We can also write
non_primes = Bloom.new(19, 100_000)
# or (it will choose the next smaller size)
non_primes = Bloom.new_by_byte_size(64 * 1024, 100_000)

# Let's put some non primes in:
Bloom.insert(non_primes, 42)
Bloom.insert(non_primes, "one hundred")
Bloom.insert(non_primes, [:ok, {Anything, 1.0e9}])

# Let's double check:
Bloom.has?(non_primes, 42)
Bloom.has?(non_primes, "one hundred")
Bloom.has?(non_primes, 7)
Bloom.has?(non_primes, [:ok, {Anything, 1.0e9}])
Bloom.has?(non_primes, 52)
# Works!


# Now we can get a estimate of
# the number of items we
# put in.
Bloom.estimate_count(non_primes)


# Apply Sieve of Eratosthenes.
# This may take a few seconds.
# At least we can do it with
# constant memory.
for x <- 2..50_000 do
  # Skip multiples of previous numbers.
  # This is actually not safe to do
  # with a bloom filter since they are
  # approximations. But let's don't care for now.
  if(Bloom.has_not?(non_primes, x)) do
    Step.throught(x*2, 100_000, x, fn non_prime ->
        # Much of the time is used to calculate
        # hashes.
        Bloom.insert(non_primes, non_prime)
    end)
 end
end && :ok

non_primes |> Bloom.has_not?(12) # not a prime
non_primes |> Bloom.has_not?(11) # prime
non_primes |> Bloom.has_not?(6719) # no prime
non_primes |> Bloom.has_not?(4245) # prime
non_primes |> Bloom.has_not?(9973) # no prime
non_primes |> Bloom.has_not?(3549) # prime
non_primes |> Bloom.has_not?(89591) # prime
non_primes |> Bloom.has_not?(84949) # no prime

# Let's write it to disk:
filename = "prime_filter.bloom"
file = File.stream!(filename, [:delayed_write, :read_ahead, :binary], 8096)
(Bloom.stream(non_primes)
|> Stream.into(file)
|> Stream.run)
# Let's inspect the size of the filter:
File.lstat!(filename).size

# We can also read from disk:
{:ok, new_non_primes} = Bloom.from_stream(file)
new_non_primes |> Bloom.has_not?(12) # not a prime
new_non_primes |> Bloom.has_not?(11) # prime
# Works!

# Maybe a bit high, 64 KB for 100_000 Numbers
# is not that much.
Bloom.false_positive_probability(non_primes)

# Let's reestimate .
# There are 9_592 primes below 100_000, so this
# should yield about 100_000 - 9_592 = 90_408.
Bloom.estimate_count(non_primes)
```
Side note: If you want to check primes, ~~google~~
search for *Miller–Rabin primality test*.


CHANGELOG.md
# Changelog

## v0.1.4

* replace `Flower.Bloom.(de)serialize` with a `Enumerable` to reduce memory usage
  when saving to disk.
  * add `Flower.Bloom.stream(bloomfilter)`
  * add `Flower.Bloom.from_stream(stream)`
* refactor dirty NIFs to be "clean"
* increase internal max hashes of Bloom Filters from 8 to 16,
  introduce `sha512`.
* include CHANGELOG

## v0.1.3

* rename `Flower.Bloom.has_maybe?` to `Flower.Bloom.has?`

## v0.1.2

* fix hex package setup

## v0.1.1

* flag `Flower.Bloom.serialize` as experimental
* flag `Flower.Bloom.deserialize` as experimental
* initial version


test/bitarray_test.exs
defmodule BitArrayTest do
  use ExUnit.Case
  alias Flower.Native.BitArray, as: BitArray
  doctest Flower.Native.BitArray

  test "1000 Random Writes And Reads" do
    a = BitArray.new(1000)

    trues = 0..999 |> Enum.to_list() |> Enum.filter(fn _ -> :rand.uniform(2) == 1 end)

    for x <- 0..999 do
      BitArray.put(a, x, x in trues)
    end

    for x <- 0..999 do
      assert BitArray.get(a, x) == x in trues
    end
  end

  test "1000 Random Writes And Reads With Serialisation" do
    a = BitArray.new(1000)

    trues = 0..999 |> Enum.to_list() |> Enum.filter(fn _ -> :rand.uniform(2) == 1 end)

    for x <- 0..999 do
      BitArray.put(a, x, x in trues)
    end

    bin = BitArray.to_bin(a)
    bs = BitArray.from_bin(bin)

    for x <- 0..999 do
      assert BitArray.get(bs, x) == x in trues
    end
  end

  test "1000 Random Writes And Reads With Stream Serialisation" do
    a = BitArray.new(1000)

    trues = 0..999 |> Enum.to_list() |> Enum.filter(fn _ -> :rand.uniform(2) == 1 end)

    for x <- 0..999 do
      BitArray.put(a, x, x in trues)
    end

    bs = BitArray.new(1000)

    BitArray.stream(a)
    |> Stream.into(BitArray.stream(bs))
    |> Stream.run()

    for x <- 0..999 do
      assert BitArray.get(bs, x) == x in trues
    end
  end

  test "1000 Random Writes And Reads With Stream To File Serialisation" do
    File.rm("test/test.file1")

    a = BitArray.new(10000)

    trues = 0..999 |> Enum.to_list() |> Enum.filter(fn _ -> :rand.uniform(2) == 1 end)

    for x <- 0..999 do
      BitArray.put(a, x, x in trues)
    end

    bs = BitArray.new(10000)

    BitArray.stream(a)
    |> Stream.into(File.stream!("test/test.file1"))
    |> Stream.run()

    File.stream!("test/test.file1")
    |> Stream.into(BitArray.stream(bs))
    |> Stream.run()

    for x <- 0..999 do
      assert BitArray.get(bs, x) == x in trues
    end
  end

  test "10_000 Random Writes True then False And Reads" do
    a = BitArray.new(10000)

    trues = 0..9999 |> Enum.to_list() |> Enum.filter(fn _ -> :rand.uniform(2) == 1 end)

    for x <- 0..9999 do
      BitArray.put(a, x, x in trues)
    end

    for x <- 0..9999 do
      BitArray.put(a, x, false)
    end

    for x <- 0..9999 do
      assert BitArray.get(a, x) == false
    end
  end

  test "Count Ones (len 10000)" do
    a = BitArray.new(10000)

    trues = 0..9999 |> Enum.to_list() |> Enum.filter(fn _ -> :rand.uniform(2) == 1 end)

    for x <- 0..9999 do
      BitArray.put(a, x, x in trues)
    end

    assert length(trues) == BitArray.count_ones(a)
  end

  test "Get Size" do
    a = BitArray.new(10000)
    assert abs(10000 - BitArray.bit_length(a)) <= 64
  end
end


test/flower_bloom_test.exs
defmodule FlowerBloomTest do
  use ExUnit.Case
  doctest Flower.Bloom

  alias Flower.Bloom, as: Bloom

  defp random_elements do
    n = :rand.uniform(1000) + 100
    0..n |> Enum.map(fn x -> 1000 * x + :rand.uniform(1000) end)
  end

  test "Test constructors" do
    non_primes_a = Bloom.new(:"64 KB", 100_000)
    non_primes_b = Bloom.new(19, 100_000)
    non_primes_c = Bloom.new_by_byte_size(64 * 1024, 100_000)

    assert Bloom.serialize(non_primes_a) == Bloom.serialize(non_primes_b)
    assert Bloom.serialize(non_primes_b) == Bloom.serialize(non_primes_c)
  end

  test "Reserialize" do
    filter = Bloom.new(:"16 KB", 10_000)

    for x <- random_elements() do
      Bloom.insert(filter, "#{x}")
      Bloom.insert(filter, x)
    end

    Bloom.insert(filter, random_elements())

    bin = Bloom.serialize(filter)
    assert bin == bin |> Bloom.deserialize() |> Bloom.serialize()
  end

  test "insert elements" do
    filter = Bloom.new(:"64 KB", 1000)

    Bloom.insert(filter, 42)
    Bloom.insert(filter, "one hundred")
    Bloom.insert(filter, [:ok, {Anything, 1.0e9}])

    assert Bloom.has_maybe?(filter, 42)
    assert Bloom.has_maybe?(filter, "one hundred")
    assert Bloom.has_not?(filter, 7)
    assert Bloom.has_maybe?(filter, [:ok, {Anything, 1.0e9}])
    assert Bloom.has_not?(filter, 52)
  end

  test "false positive probability" do
    filter = Bloom.new(:"64 KB", 10_000)

    for x <- 1..10_000 do
      Bloom.insert(filter, x)
    end

    real_false_positives =
      (for x <- 1..20_000 do
         if Bloom.has_maybe?(filter, x) do
           1
         else
           0
         end
       end
       |> Enum.sum()) - 10_000

    calc_false_positives = 10_000 * Bloom.false_positive_probability(filter)

    # 0.1% should be okay
    assert abs(real_false_positives - calc_false_positives) < 10
  end

  test "estimate count" do
    filter = Bloom.new(:"8 KB", 1000)

    for x <- 1..1_000 do
      Bloom.insert(filter, x)
    end

    # 1% should be okay
    assert Bloom.estimate_count(filter) - 1_000 < 10
  end

  test "stream into file and back" do
    File.rm("test/test.file3")

    filter = Bloom.new(:"128 KB", 1000)

    for x <- 1..1_000 do
      Bloom.insert(filter, "#{x}")
    end

    Bloom.stream(filter)
    |> Stream.into(File.stream!("test/test.file3", [], 1024))
    |> Stream.run()

    {:ok, filter2} = Bloom.from_stream(File.stream!("test/test.file3", [], 1024))

    assert Bloom.serialize(filter) == Bloom.serialize(filter2)
  end

  test "copy with stream" do
    filter = Bloom.new(:"512 KB", 1000)

    Bloom.insert(filter, true)

    assert Bloom.has?(filter, true) == true
    assert Bloom.has?(filter, false) == false

    {:ok, new_filter} =
      filter
      |> Bloom.stream()
      |> Bloom.from_stream()

    assert Bloom.has?(new_filter, true) == true
    assert Bloom.has?(new_filter, false) == false
  end

  # This test takes too long then not run in production mode.
  if Mix.env() == :prod do
    test "stream into file and back 512MB" do
      File.rm("test/test.file2")

      filter = Bloom.new(:"512 MB", 1000)

      for x <- 1..1_000 do
        Bloom.insert(filter, "#{x}")
      end

      me = self()

      spawn_link(fn ->
        send(me, Bloom.stream(filter) |> hash_stream)
      end)

      Bloom.stream(filter)
      |> Stream.into(
        File.stream!(
          "test/test.file2",
          [:delayed_write, :read_ahead, :binary, :compressed],
          64 * 1024
        )
      )
      |> Stream.run()

      # Note: :compressed is only used because the Bloom filter is near empty
      {:ok, filter2} =
        Bloom.from_stream(
          File.stream!(
            "test/test.file2",
            [:delayed_write, :read_ahead, :binary, :compressed],
            64 * 1024
          )
        )

      spawn_link(fn ->
        send(me, Bloom.stream(filter2) |> hash_stream)
      end)

      hash_one =
        receive do
          a -> a
        after
          60_000 ->
            throw("failed")
        end

      hash_two =
        receive do
          a -> a
        after
          60_000 ->
            throw("failed")
        end

      assert hash_one == hash_two
    end

    defp hash_stream(stream) do
      stream
      |> Enum.reduce(:crypto.hash_init(:md5), fn bin, acc ->
        :crypto.hash_update(acc, bin)
      end)
      |> :crypto.hash_final()
    end
  end
end


mix.exs
defmodule Flower.MixProject do
  use Mix.Project

  def project do
    [
      app: :flower,
      name: "flower",
      version: "0.1.4",
      elixir: "~> 1.6",
      build_embedded: Mix.env() == :prod,
      start_permanent: Mix.env() == :prod,
      compilers: [:rustler] ++ Mix.compilers(),
      rustler_crates: rustler_crates(),
      package: package(),
      description: description(),
      deps: deps(),
      source_url: "https://github.com/CodeSteak/Flower",
      docs: [
        main: "readme",
        extras: ["README.md"]
      ]
    ]
  end

  defp package() do
    [
      maintainers: ["Codesteak"],
      licenses: ["MIT"],
      links: %{
        "Github" => "https://github.com/CodeSteak/Flower"
      },
      files: [
        "mix.exs",
        "native/bitarray/src",
        "native/bitarray/Cargo.toml",
        "native/bitarray/TODO.md",
        "lib",
        "LICENSE",
        "README.md",
        "CHANGELOG.md"
      ]
    ]
  end

  def description() do
    """
    This is an implementation of Bloom Filters for Elixir. It uses Rust NIFs for better performance.
    """
  end

  # Run "mix help compile.app" to learn about applications.
  def application do
    [
      extra_applications: [:logger]
    ]
  end

  # Run "mix help deps" to learn about dependencies.
  defp deps do
    [
      {:rustler, "~> 0.16.0"},
      {:ex_doc, "~> 0.16", only: :dev, runtime: false}
    ]
  end

  defp rustler_crates do
    [
      bitarray: [
        path: "native/bitarray",
        mode: rustc_mode(Mix.env())
      ]
    ]
  end

  defp rustc_mode(:prod), do: :release
  defp rustc_mode(_), do: :debug
end


lib/flower/native.ex
defmodule Flower.Native.BitArray do
  @moduledoc false

  use Rustler, otp_app: :flower, crate: :bitarray

  def new(_), do: :erlang.nif_error(:nif_not_loaded)
  def put(_, _, _), do: :erlang.nif_error(:nif_not_loaded)
  def get(_, _), do: :erlang.nif_error(:nif_not_loaded)
  def to_bin_chunked(_, _), do: :erlang.nif_error(:nif_not_loaded)
  def or_chunk(_, _, _), do: :erlang.nif_error(:nif_not_loaded)
  def bit_length(_), do: :erlang.nif_error(:nif_not_loaded)
  def count_ones_chunked(_, _), do: :erlang.nif_error(:nif_not_loaded)

  def count_ones(ref) do
    count_ones(ref, {0, 0})
  end

  def count_ones(_ref, {:eof, sum}) do
    sum
  end

  def count_ones(ref, {next_chunk, sum}) do
    sum + count_ones(ref, count_ones_chunked(ref, next_chunk))
  end

  def to_bin(ref) do
    to_bin(ref, {0, <<>>})
  end

  def to_bin(_ref, {:eof, data}) do
    data
  end

  def to_bin(ref, {next_chunk, data}) do
    data <>
      to_bin(
        ref,
        to_bin_chunked(ref, next_chunk)
      )
  end

  def from_bin(bin) do
    ref =
      bin
      |> bit_size()
      |> new()

    ref
    |> or_chunk(bin, 0)

    ref
  end
end


lib/flower/native_bitarray.ex
defmodule Flower.Native.BitArray do
  @moduledoc false

  use Rustler, otp_app: :flower, crate: :bitarray

  def new(_), do: :erlang.nif_error(:nif_not_loaded)
  def put(_, _, _), do: :erlang.nif_error(:nif_not_loaded)
  def get(_, _), do: :erlang.nif_error(:nif_not_loaded)
  def to_bin_chunked(_, _), do: :erlang.nif_error(:nif_not_loaded)
  def or_chunk(_, _, _), do: :erlang.nif_error(:nif_not_loaded)
  def bit_length(_), do: :erlang.nif_error(:nif_not_loaded)
  def count_ones_chunked(_, _), do: :erlang.nif_error(:nif_not_loaded)

  def count_ones(ref) do
    count_ones(ref, {0, 0})
  end

  def count_ones(_ref, {:eof, sum}) do
    sum
  end

  def count_ones(ref, {next_chunk, sum}) do
    sum + count_ones(ref, count_ones_chunked(ref, next_chunk))
  end

  def to_bin(ref) do
    to_bin(ref, {0, <<>>})
  end

  def to_bin(_ref, {:eof, data}) do
    data
  end

  def to_bin(ref, {next_chunk, data}) do
    data <>
      to_bin(
        ref,
        to_bin_chunked(ref, next_chunk)
      )
  end

  def from_bin(bin) do
    ref =
      bin
      |> bit_size()
      |> new()

    ref
    |> or_chunk(bin, 0)

    ref
  end

  def stream(ref) do
    %Flower.Native.BitArray.Stream{bitarray_ref: ref, pos: 0}
  end
end


lib/flower/native_bitarray_stream.ex
defmodule Flower.Native.BitArray.Stream do
  @moduledoc false
  defstruct bitarray_ref: nil, pos: 0

  alias Flower.Native.BitArray, as: BitArray

  defimpl Enumerable do
    # def reduce(_,       {:halt, acc}, _fun),   do: {:halted, acc}
    # def reduce(list,    {:suspend, acc}, fun), do: {:suspended, acc, &reduce(list, &1, fun)}
    # def reduce([],      {:cont, acc}, _fun),   do: {:done, acc}
    # def reduce([h | t], {:cont, acc}, fun),    do: reduce(t, fun.(h, acc), fun)

    def reduce(%{bitarray_ref: _ref, pos: _pos}, {:halt, acc}, _fun), do: {:halted, acc}

    def reduce(%{bitarray_ref: _ref, pos: _pos} = stream, {:suspend, acc}, fun),
      do: {:suspended, acc, &reduce(stream, &1, fun)}

    def reduce(%{bitarray_ref: _ref, pos: :eof}, {:cont, acc}, _fun), do: {:done, acc}

    def reduce(%{bitarray_ref: ref, pos: pos}, {:cont, acc}, fun) do
      {nextpos, data} = BitArray.to_bin_chunked(ref, pos)

      reduce(%{bitarray_ref: ref, pos: nextpos}, fun.(data, acc), fun)
    end

    def count(_stream) do
      {:error, __MODULE__}
    end

    def member?(_stream, _term) do
      {:error, __MODULE__}
    end

    def slice(_stream) do
      {:error, __MODULE__}
    end
  end

  defimpl Collectable do
    def into(%{bitarray_ref: _ref, pos: 0} = stream) do
      {stream, &collector_func/2}
    end

    defp collector_func(%{bitarray_ref: ref, pos: offset}, {:cont, elm}) when is_binary(elm) do
      next_offset = BitArray.or_chunk(ref, elm, offset)
      %{bitarray_ref: ref, pos: next_offset}
    end

    defp collector_func(%{bitarray_ref: ref, pos: _offset}, :done) do
      ref
    end

    defp collector_func(%{bitarray_ref: _ref, pos: _offset}, :halt) do
      :ok
    end
  end
end


lib/flower/bloom.ex
defmodule Flower.Bloom do
  use Bitwise
  alias Flower.Native.BitArray, as: BitArray

  @moduledoc """
  Flower.Bloom implements a Bloom Filter.

  For this Bloom Filter sha256 or sha512 is used as hash function.
  """

  @ser_vsn 1

  @byte_sizes [
    :"8 Byte",
    :"16 Byte",
    :"32 Byte",
    :"64 Byte",
    :"128 Byte",
    :"256 Byte",
    :"512 Byte",
    :"1 KB",
    :"2 KB",
    :"4 KB",
    :"8 KB",
    :"16 KB",
    :"32 KB",
    :"64 KB",
    :"128 KB",
    :"256 KB",
    :"512 KB",
    :"1 MB",
    :"2 MB",
    :"4 MB",
    :"8 MB",
    :"16 MB",
    :"32 MB",
    :"64 MB",
    :"128 MB",
    :"256 MB",
    :"512 MB"
  ]

  @type bloomfilter ::
          {:bloom, bitarray :: reference(), bitaddrmask :: integer(), number_of_hashes :: 1..8}
  @type size_atom ::
          :"8 Byte"
          | :"16 Byte"
          | :"32 Byte"
          | :"64 Byte"
          | :"128 Byte"
          | :"256 Byte"
          | :"512 Byte"
          | :"1 KB"
          | :"2 KB"
          | :"4 KB"
          | :"8 KB"
          | :"16 KB"
          | :"32 KB"
          | :"64 KB"
          | :"128 KB"
          | :"256 KB"
          | :"512 KB"
          | :"1 MB"
          | :"2 MB"
          | :"4 MB"
          | :"8 MB"
          | :"16 MB"
          | :"32 MB"
          | :"64 MB"
          | :"128 MB"
          | :"256 MB"
          | :"512 MB"

  @doc """
  Create a new Bloom Filter with `size` :: `size_atom()` or 2^bitaddrlen bits.

  |bitaddrlen|    Size|Bitaddrlen|    Size|Bitaddrlen|    Size|
  |-------:|---------:|-------:|---------:|-------:|---------:|
  | __  __ |          | __13__ | 1 KB     | __23__ | 1 MB     |
  | __  __ |          | __14__ | 2 KB     | __24__ | 2 MB     |
  | __  __ |          | __15__ | 4 KB     | __25__ | 4 MB     |
  | __6 __ | 8 Byte   | __16__ | 8 KB     | __26__ | 8 MB     |
  | __7 __ | 16 Byte  | __17__ | 16 KB    | __27__ | 16 MB    |
  | __8 __ | 32 Byte  | __18__ | 32 KB    | __28__ | 32 MB    |
  | __9 __ | 64 Byte  | __19__ | 64 KB    | __29__ | 64 MB    |
  | __10__ | 128 Byte | __20__ | 128 KB   | __30__ | 128 MB   |
  | __11__ | 256 Byte | __21__ | 256 KB   | __31__ | 256 MB   |
  | __12__ | 512 Byte | __22__ | 512 KB   | __32__ | 512 MB   |
  """
  @spec new(bitaddrlen :: 6..32, expected_elements :: pos_integer()) :: bloomfilter()
  def new(bitaddrlen, expected_elements) when bitaddrlen in 6..32 do
    number_of_hashes =
      1..16 |> Enum.min_by(&calc_fp_prob(expected_elements, 1 <<< bitaddrlen, &1))

    if number_of_hashes == 1 do
      IO.warn("Your Bloom filter is too small for the expected number of elements!")
    end

    {:bloom, BitArray.new(1 <<< bitaddrlen), (1 <<< bitaddrlen) - 1, number_of_hashes}
  end

  @spec new(bitaddrlen :: size_atom(), expected_elements :: pos_integer()) :: bloomfilter()
  def new(bytes, expected_elements) when bytes in @byte_sizes do
    bitaddrlen = 6 + Enum.find_index(@byte_sizes, fn x -> x == bytes end)
    new(bitaddrlen, expected_elements)
  end

  @doc """
  Create a new Bloom Filter with maximum byte size 'bytes'. The size gets
  rounded down to the next `size_atom()`.
  """
  @spec new_by_byte_size(bytes :: size_atom(), expected_elements :: pos_integer()) ::
          bloomfilter()
  def new_by_byte_size(bytes, expected_elements) when bytes in @byte_sizes do
    new(bytes, expected_elements)
  end

  @spec new_by_byte_size(bytes :: pos_integer(), expected_elements :: pos_integer()) ::
          bloomfilter()
  def new_by_byte_size(bytes, expected_elements) do
    bitaddrlen = trunc(:math.log2(bytes * 8))

    new(bitaddrlen, expected_elements)
  end

  defp calc_fp_prob(elem, size, number_of_hashes) do
    e = 2.71828182846
    fraction_of_0 = :math.pow(e, -number_of_hashes * elem / size)
    fraction_of_1 = 1 - fraction_of_0

    false_positives = :math.pow(fraction_of_1, number_of_hashes)
    false_positives
  end

  @doc """
  Calculates the false positive probability for a given bloom filter.
  Return value is between `0.0` and `1.0`.
  This Operation is slow for large Bloom Filters and should then be avoided.
  """
  @spec false_positive_probability(bloomfilter()) :: float()
  def false_positive_probability({:bloom, bitarray, _mask, number_of_hashes}) do
    fraction_of_1 = BitArray.count_ones(bitarray) / BitArray.bit_length(bitarray)
    false_positives = :math.pow(fraction_of_1, number_of_hashes)

    false_positives
  end

  @doc """
  Estimates how many unique elements have been added.
  This Operation is slow for large Bloom Filters and should then be avoided.
  """
  @spec estimate_count(bloomfilter()) :: non_neg_integer()
  def estimate_count({:bloom, bitarray, _mask, number_of_hashes}) do
    bits = BitArray.bit_length(bitarray)
    ones = BitArray.count_ones(bitarray)

    fraction_of_1 = ones / bits
    fraction_of_0 = 1 - fraction_of_1

    elmements = -1 * :math.log(fraction_of_0) * bits / number_of_hashes

    round(elmements)
  end

  @doc """
  Inserts an Erlang Term into the Bloom Filter.
  """
  @spec insert(bloomfilter(), any()) :: :ok
  def insert({:bloom, bitarray, mask, number_of_hashes}, bin) when is_binary(bin) do
    bin
    |> bin_to_offset_list(number_of_hashes)
    |> Enum.map(&Bitwise.&&&(&1, mask))
    |> write(bitarray)
  end

  def insert(bloom, term) do
    insert(bloom, :erlang.term_to_binary(term))
  end

  @doc """
  Checks if an element was inserted in the given Bloom Filter.

  Returns `false` if it can be guaranteed that the element was not
  inserted. Else `true`.

  You can get the probability of a false positive
  using `Flower.Bloom.false_positive_probability`.

  |Was actually inserted?|                  has?  |                    has_not?  |
  |:--------------------:|:----------------------:|:----------------------------:|
  | yes                  | yes                    |                       no     |
  | no                   | most of the times: no  |   most of the times: yes     |

  """
  @spec has?(bloomfilter(), any()) :: boolean()
  def has?({:bloom, bitarray, mask, number_of_hashes}, bin) when is_binary(bin) do
    bin
    |> bin_to_offset_list(number_of_hashes)
    |> Enum.map(&Bitwise.&&&(&1, mask))
    |> read(bitarray)
  end

  def has?(bloom, term) do
    has?(bloom, :erlang.term_to_binary(term))
  end

  @doc """
  Checks if an element is not in a given Bloom Filter.

  Returns `true` if it can be guaranteed that the element was not
  inserted. Else `false`.

  This is equal to `!Bloom.has?(filter, term)`
  """
  @spec has_not?(bloomfilter(), any()) :: boolean()
  def has_not?(bloom, term), do: !has?(bloom, term)

  @doc false
  @deprecated "This is unstable, can change soon"
  def serialize({:bloom, bitarray, _mask, number_of_hashes}) do
    blen = BitArray.bit_length(bitarray)
    bitaddrlen = :math.log2(blen) |> trunc()
    <<@ser_vsn, 42, bitaddrlen::8, number_of_hashes::8, BitArray.to_bin(bitarray)::binary>>
  end

  @doc """
  Creates a `Stream` of binaries. This should be used for serializing.

  Example:
  ```
  filter = Bloom.new(...)
  file   = File.stream!("myfile.bloomfilter", [:delayed_write, :binary], 8096)

  Bloom.stream(filter)
  |> Stream.into(file)
  |> Stream.run
  ```
  """
  @spec stream(bloomfilter()) :: Enumerable.t()
  def stream({:bloom, bitarray, _mask, number_of_hashes}) do
    blen = BitArray.bit_length(bitarray)
    bitaddrlen = :math.log2(blen) |> trunc()

    [<<@ser_vsn, 42, bitaddrlen::8, number_of_hashes::8>>]
    |> Stream.concat(BitArray.stream(bitarray))
  end

  @doc false
  @deprecated "This is unstable, can change soon"
  def deserialize(<<@ser_vsn, 42, bitaddrlen::8, number_of_hashes::8, bitarray::binary>>) do
    {:bloom, BitArray.from_bin(bitarray), (1 <<< bitaddrlen) - 1, number_of_hashes}
  end

  @doc """
  Creates a Bloom Filter from a `Stream` of binaries. This should be used for deserializing.

  Example:
  ```
  file   = File.stream!("myfile.bloomfilter", [:read_ahead, :binary], 8096)
  Bloom.from_stream(file)
  ```
  """
  @spec from_stream(Enumerable.t()) :: {:ok, bloomfilter()} | {:error, any()}
  def from_stream(stream), do: from_stream(stream, <<>>)

  defp from_stream(stream, <<@ser_vsn, 42, bitaddrlen::8, number_of_hashes::8, tail::binary>>) do
    ref = BitArray.new(1 <<< bitaddrlen)

    [tail]
    |> Stream.concat(stream)
    |> Stream.into(BitArray.stream(ref))
    |> Stream.run()

    {:ok, {:bloom, ref, (1 <<< bitaddrlen) - 1, number_of_hashes}}
  end

  defp from_stream(stream, acc) when byte_size(acc) < 100 do
    [head] = stream |> Enum.take(1)
    next_acc = acc <> head
    next_stream = stream |> Stream.drop(1)
    from_stream(next_stream, next_acc)
  end

  defp from_stream(_, _) do
    {:error, :invalid_header}
  end

  defp write([p | tail], bitarray) do
    BitArray.put(bitarray, p, true)
    write(tail, bitarray)
  end

  defp write([], _) do
    :ok
  end

  defp read([p | tail], bitarray) do
    BitArray.get(bitarray, p) && read(tail, bitarray)
  end

  defp read([], _) do
    true
  end

  defp bin_to_offset_list(bin, number_of_hashes) when number_of_hashes <= 8 do
    :crypto.hash(:sha256, bin)
    |> hash_to_offset_list(number_of_hashes)
  end

  defp bin_to_offset_list(bin, number_of_hashes) when number_of_hashes > 8 do
    :crypto.hash(:sha512, bin)
    |> hash_to_offset_list(number_of_hashes)
  end

  defp hash_to_offset_list(_, 0), do: []

  defp hash_to_offset_list(<<num::32, rest::binary>>, n),
    do: [num | hash_to_offset_list(rest, n - 1)]

  @doc false
  @deprecated "Use `has?` instead"
  def has_maybe?(a, b), do: has?(a, b)
end
